### PR TITLE
implied bounds: normalize in the proper param_env

### LINKED ISSUE #95922: TAIT regression: requires unnecessary explicit lifetime bound
I tried this code:

```rust
#![feature(type_alias_impl_trait)]
use core::future::Future;

trait Service<'a, Req> {
    type Future: Future;
    fn call(req: &'a Req) -> Self::Future;
}

impl<'a, Req> Service<'a, Req> for u8 {
    type Future = impl Future;
    fn call(req: &'a Req) -> Self::Future {
        async move { let x = req; }
    }
}
```

([Playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=3b4c470458a245ff1f8a032101641de2))

This code used to compile fine prior to #95519. It now requires an explicit lifetime bound `Req: 'a` to the impl.

Unfortunately, adding a lifetime bound may not be an option when used with HRTBS due to bugs like #95921. This caused a major breakage on my part.

### Meta

regressed in the nightly version `1.62.0-nightly (f4a7ce997 2022-04-08)`

<details><summary>Error Output</summary>
<p>

```
   Compiling rust-ci v0.1.0 (/home/runner/work/rust-ci/rust-ci)
error[E0309]: the parameter type `Req` may not live long enough
Error:   --> src/lib.rs:10:19
   |
10 |     type Future = impl Future;
   |                   ^^^^^^^^^^^
   |
   = help: consider adding an explicit lifetime bound `Req: 'a`...
   = note: ...so that the reference type `&'a Req` does not outlive the data it points at

For more information about this error, try `rustc --explain E0309`.
error: could not compile `rust-ci` due to previous error
```

</p>
</details>

### PR DESCRIPTION
**Pull request by aliemjay**
TL;DR This moves the normalization of implied wf types from `ObligationCtxt::assumed_wf_types` to `TyCtxt::assumed_wf_types` query and also implements #105779 for the query.

Using different ParamEnv's when normalizng wf types can lead to different results and is more error-prone. Now we are only using the original item's param env. I can't think of a way this changes user-facing behavior, only because #98852 isn't fixed yet.

The also uses early and late binders for `assumed_wf_types` query. The motivation was cleanly fix #95922, which needs to substitute early-bound params for the defined opaque type.

r? @lcnr

### CODE REVIEW

File: compiler/rustc_ty_utils/src/implied_bounds.rs [hunk 1]
```diff
@@ -56,6 +67,38 @@ fn assumed_wf_types<'tcx>(tcx: TyCtxt<'tcx>, def_id: DefId) -> &'tcx ty::List<Ty
...
+        | DefKind::Generator => ty::EarlyBinder(ty::Binder::dummy(ty::List::empty())),
     }
 }
+
+fn from_fn_sig<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    def_id: DefId,
+) -> ty::Binder<'tcx, &'tcx ty::List<Ty<'tcx>>> {
+    // FIXME(#84533): We probably shouldn't use output types in implied bounds.
+    // This would reject this fn `fn f<'a, 'b>() -> &'a &'b () { .. }`.
```
Comment (lcnr) [inline compiler/rustc_ty_utils/src/implied_bounds.rs:79]: I don't think this should be how we fix #84533 because you'd still have an builtin impl with an associated type which isn't well formed.

The correct fix is to actually check that the return type is well formed even without calling the function. This needs implications for binders though.

Until then something like #106807 might work as a stopgap (again having the issue of unnormalized projections though)

File: compiler/rustc_ty_utils/src/implied_bounds.rs [hunk 2]
```diff
@@ -56,6 +67,38 @@ fn assumed_wf_types<'tcx>(tcx: TyCtxt<'tcx>, def_id: DefId) -> &'tcx ty::List<Ty
...
+    value: T,
+) -> Result<T, NoSolution> {
+    use rustc_infer::infer::TyCtxtInferExt as _;
+    use rustc_infer::traits::ObligationCause;
+    use rustc_trait_selection::traits::query::normalize::QueryNormalizeExt as _;
+
+    let infcx = tcx.infer_ctxt().build();
+    let normalized = infcx
+        .at(&ObligationCause::dummy(), param_env)
+        .query_normalize(value)
```
Comment (lcnr) [inline compiler/rustc_ty_utils/src/implied_bounds.rs:100]: `query_normalize` eagerly evaluates constants causing cycle errors if there are unevaluated constants in the fn sig.

During analysis `normalize` should be used instead.

**Review by lcnr — COMMENTED** [review]
nits, then r=me

**Comment by apiraino** [comment]
Switching to waiting on author for a few changes, then it looks we're ok here?

@rustbot author

**Comment by JohnCSimon** [comment]
@aliemjay  Ping from triage:  Can you post your status on this PR? This has sat idle for a few months.

**Comment by Dylan-DPC** [comment]
Closing this as inactive. Feel free to reöpen this pr or create a new pr if you get the time to work on this. Thanks
